### PR TITLE
Individual widget fonts

### PR DESCRIPTION
**Pull request by N1x0**
First step to enable icon fonts as mentioned in #152 . This PR allows each widget to set its own font and effectively overwrite the default font of the bar.

### CODE REVIEW

File: src/workspacer.Bar/Widgets/ActiveLayoutWidget.cs [hunk 1]
```diff
@@ -9,6 +9,7 @@ namespace workspacer.Bar.Widgets
 {
     public class ActiveLayoutWidget : BarWidgetBase
     {
+        public string FontName { get; set; } = null;
```
Comment (josteink) [inline src/workspacer.Bar/Widgets/ActiveLayoutWidget.cs:12]: Any reason not to just put this in BarWidgetBase, if all widgets are going to have this?

File: src/workspacer.Bar/Widgets/BatteryWidget.cs [hunk 2]
```diff
@@ -17,6 +17,7 @@ public class BatteryWidget : BarWidgetBase
         public double LowChargeThreshold { get; set; } = 0.10;
         public double MedChargeThreshold { get; set; } = 0.50;
         public int Interval { get; set; } = 5000;
+        public string FontName { get; set; } = null;
```
Comment (josteink) [inline src/workspacer.Bar/Widgets/BatteryWidget.cs:20]: Same about this declaration. Why not just inherit it from base instead?

File: src/workspacer.Bar/Widgets/BatteryWidget.cs [hunk 3]
```diff
@@ -29,20 +30,20 @@ public override IBarWidgetPart[] GetParts()
             {
                 if (currentBatteryCharge <= LowChargeThreshold)
                 {
-                    return Parts(Part(currentBatteryCharge.ToString("#0%"), LowChargeColor));
+                    return Parts(Part(currentBatteryCharge.ToString("#0%"), LowChargeColor, null, null, FontName));
```
Comment (josteink) [inline src/workspacer.Bar/Widgets/BatteryWidget.cs:33]: These invocation might look cleaner if using [named arguments](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/named-and-optional-arguments).

File: src/workspacer.Bar/Widgets/FocusedMonitorWidget.cs [hunk 4]
```diff
@@ -8,14 +8,16 @@ namespace workspacer.Bar.Widgets
 {
     public class FocusedMonitorWidget : BarWidgetBase
     {
+        public string FontName { get; set; } = null;
```
Comment (josteink) [inline src/workspacer.Bar/Widgets/FocusedMonitorWidget.cs:11]: This could be inherited too.

File: src/workspacer.Bar/Widgets/TextWidget.cs [hunk 5]
```diff
@@ -8,6 +8,7 @@ namespace workspacer.Bar.Widgets
 {
     public class TextWidget : BarWidgetBase
     {
+        public string FontName { get; set; } = null;
```
Comment (josteink) [inline src/workspacer.Bar/Widgets/TextWidget.cs:11]: And this one.

File: src/workspacer.Bar/Widgets/TextWidget.cs [hunk 6]
```diff
@@ -17,7 +18,7 @@ public TextWidget(string text)
 
         public override IBarWidgetPart[] GetParts()
         {
-            return Parts(_text);
+            return Parts(Part(_text,null,null,null,FontName));
```
Comment (josteink) [inline src/workspacer.Bar/Widgets/TextWidget.cs:21]: This invocation would *definitely* look nicer with named arguments.

File: src/workspacer.Bar/Widgets/TimeWidget.cs [hunk 7]
```diff
@@ -9,6 +9,7 @@ namespace workspacer.Bar.Widgets
 {
     public class TimeWidget : BarWidgetBase
     {
+        public string FontName { get; set; } = null;
```
Comment (josteink) [inline src/workspacer.Bar/Widgets/TimeWidget.cs:12]: Inherit this one from BarWidgetBase too.

**Comment by josteink** [comment]
I've taken a quick look and it mostly looks OK to me, but there's a bit code-repetition which can be removed, and there's some function-invocations which does look a bit risky, which IMO can also be improved.

Fix that and I'll take a second look :smiley:

File: src/workspacer.Bar/Widgets/ActiveLayoutWidget.cs (hunk 1 above)
Comment (N1x0) [reply src/workspacer.Bar/Widgets/ActiveLayoutWidget.cs]: Blanked on the fact that this was possible... Thanks for pointing it out!

**Comment by N1x0** [comment]
Updated based on your feedback 😊

File: src/workspacer.Bar/Widgets/TextWidget.cs [hunk 8]
```diff
@@ -18,7 +18,7 @@ public TextWidget(string text)
 
         public override IBarWidgetPart[] GetParts()
         {
-            return Parts(Part(_text,null,null,null,FontName));
+            return Parts(Part(_text, fore: null, back: null, partClicked: null, FontName));
```
Comment (josteink) [inline src/workspacer.Bar/Widgets/TextWidget.cs:21]: IN this case I was more thinking eliminating redundant parameters using named arguments. Like this:

````c#
return Parts(Part(_text, fontName: FontName));
````

File: src/workspacer.Bar/Widgets/WorkspaceWidget.cs [hunk 9]
```diff
@@ -14,7 +14,7 @@ public class WorkspaceWidget : BarWidgetBase
         public Color WorkspaceEmptyColor { get; set; } = Color.Gray;
         public Color WorkspaceIndicatingBackColor { get; set; } = Color.Teal;
         public int BlinkPeriod { get; set; } = 1000;
-        public string FontName { get; set; } = null;
+        //public string FontName { get; set; }
```
Comment (josteink) [inline src/workspacer.Bar/Widgets/WorkspaceWidget.cs:17]: This line should be removed, not commented out.

File: src/workspacer.Bar/Widgets/TextWidget.cs (hunk 8 above)
Comment (N1x0) [reply src/workspacer.Bar/Widgets/TextWidget.cs:21]: oh my bad, totally misunderstood you there

**Comment by josteink** [comment]
That looks much better. Thanks!￼

Consider it merged!￼

**Review by josteink — APPROVED** [review]

**Pull request merged**